File: wrappers/javascript/shared/src/store/EntryList.ts
import type { EntryObject } from '.'
import type { EntryListHandle } from '../crypto'

import { ariesAskar } from '../ariesAskar'

import { Entry } from './Entry'

export class EntryList {
  private _handle: EntryListHandle
  private _length = 0

  public constructor({ handle, length }: { handle: EntryListHandle; length?: number }) {
    this._handle = handle
    this._length = length || ariesAskar.entryListCount({ entryListHandle: handle })
  }

  public get handle() {
    return this._handle
  }

  public get length() {
    return this._length
  }

  public getEntryByIndex(index: number) {
    return new Entry({ list: this.handle, position: index })
  }

  private forEach(cb: (entry: Entry, index?: number) => unknown) {
    for (let i = 0; i < this.length; i++) {
      cb(this.getEntryByIndex(i), i)
    }
  }

  public find(cb: (entry: Entry, index?: number) => boolean): Entry | undefined {
    for (let i = 0; i < this.length; i++) {
      const item = this.getEntryByIndex(i)
      if (cb(item)) return item
    }
  }

  public toArray(): Array<EntryObject> {
    const list: Array<EntryObject> = []
    this.forEach((entry) => list.push(entry.toJson()))
    return list
  }
}

File: wrappers/kotlin/src/commonMain/kotlin/Store/EntryList.kt
package askar.Store

import askar.Askar
import askar.crypto.EntryListHandle

class EntryList(private val handle: EntryListHandle, length: Int? = null) {
    private val length: Int

    init {
        this.length = length?: Askar.entryList.entryListCount(handle.handle)
    }

    fun handle(): EntryListHandle {
        return handle
    }

    fun length(): Int {
        return length
    }

    fun getEntryByIndex(index: Int): Entry {
        return Entry(this.handle, index)
    }

    private fun forEach(cb: (entry: Entry, index: Int) -> Any){
        for(i in 0 until this.length){
            cb(getEntryByIndex(i), i)
        }
    }

    fun find(cb: (entry: Entry, index: Int) -> Boolean): Entry? {
        for(i in 0 until this.length){
            if(cb(this.getEntryByIndex(i), i))
                return this.getEntryByIndex(i)
        }
        return null
    }

    fun toArray(): ArrayList<EntryObject> {
        val list = ArrayList<EntryObject>(this.length)
        this.forEach { entry, _ -> list.add(entry.toJson()) }
        return list
    }

}